Sources/TuistSigning/ProvisioningProfile/ProvisioningProfile.swift

```diff
@@ -0,0 +1,98 @@
+import Foundation
+import TSCBasic
+
+/// Model of a provisioning profile
+struct ProvisioningProfile: Equatable {
+    /// Path to the provisioning profile
+    var path: AbsolutePath?
+    let name: String
+    let targetName: String
+    let configurationName: String
+    let uuid: String
+    let teamId: String
+    let appId: String
+    let appIdName: String
+    let applicationIdPrefix: [String]
+    let platforms: [String]
+    let expirationDate: Date
+
+    init(path: AbsolutePath? = nil,
+         name: String,
+         targetName: String,
+         configurationName: String,
+         uuid: String,
+         teamId: String,
+         appId: String,
+         appIdName: String,
+         applicationIdPrefix: [String],
+         platforms: [String],
+         expirationDate: Date) {
+        self.path = path
+        self.name = name
+        self.targetName = targetName
+        self.configurationName = configurationName
+        self.uuid = uuid
+        self.teamId = teamId
+        self.appId = appId
+        self.appIdName = appIdName
+        self.applicationIdPrefix = applicationIdPrefix
+        self.platforms = platforms
+        self.expirationDate = expirationDate
+    }
+}
+
+extension ProvisioningProfile: Decodable {
+    private struct Entitlements: Decodable {
+        let appId: String
+
+        private enum CodingKeys: String, CodingKey {
+            case appId = "application-identifier"
+        }
+
+        init(from decoder: Decoder) throws {
+            let container = try decoder.container(keyedBy: CodingKeys.self)
+            appId = try container.decode(String.self, forKey: .appId)
+        }
+    }
+
+    private enum CodingKeys: String, CodingKey {
+        case name = "Name"
+        case uuid = "UUID"
+        case teamIds = "TeamIdentifier"
+        case appIdName = "AppIDName"
+        case applicationIdPrefix = "ApplicationIdentifierPrefix"
+        case platforms = "Platform"
+        case entitlements = "Entitlements"
+        case expirationDate = "ExpirationDate"
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        name = try container.decode(String.self, forKey: .name)
+        uuid = try container.decode(String.self, forKey: .uuid)
+        let teamIds = try container.decode([String].self, forKey: .teamIds)
+        guard
+            let teamId = teamIds.first
+        else { throw DecodingError.valueNotFound(String.self, DecodingError.Context(codingPath: container.codingPath,
+                                                                                    debugDescription: "Array of teamID must be non-empty"))
+        }
```

Comment:
Some of this custom decoder can be abstracted into types we can reuse elsewhere. e.g.

```swift
@propertyWrapper
struct DecodingFirst<Value>: Codable where Value: Codable {
    
    public var wrappedValue: Value

    public init(wrappedValue: Value, key: AnyCodingKey) {
        self.wrappedValue = wrappedValue
    }
    
    public init(from decoder: Decoder) throws {
        guard let value = try [Value](from: decoder).first else {
            throw "Expected an array"
        }
        self.wrappedValue = value
    }
    
    public func encode(to encoder: Encoder) throws {
        try [wrappedValue].encode(to: encoder)
    }
    
}
```

```swift
...
@DecodingFirst var teamId: String
...
```

Property Wrappers are a great mechanism to allow us to intercept the coding behaviour of properties. I think you with a few more specialised types you can get the synthesised code rather than the custom implementation. I'd prefer if we did that because we could start to reuse them elsewhere we want the same behaviour

Reply:
I have kept the custom decoded since I wanted a custom decoding error for decoding name and its components (which is specific for provisioning profile)
But I have also added `DecodingFirst` wrapper that is leveraged now in the custom decoding initializer